File: tree/leetcode_Binary_Tree_Inorder_Traversal.java
/*Binary_Tree_Inorder_Traversal

Given a binary tree, return the inorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},

return [1,3,2].

Note: Recursive solution is trivial, could you do it iteratively?
*/

import java.util.ArrayList;
import java.util.Stack;


public class leetcode_Binary_Tree_Inorder_Traversal {

    public static void main(String[] args) {

    }

    public static ArrayList<Integer> inorderTraversal(TreeNode root) {
        ArrayList<Integer> all = new ArrayList<Integer>();
        Stack<TreeNode> s = new Stack<TreeNode>();
        TreeNode node = root;
        while (node != null || !s.isEmpty()) {
            if (node != null) {
                s.add(node);
                node = node.left;
            } else {
                node = s.pop();
                all.add(node.val);
                node = node.right;
            }
        }
        return all;
    }

    // Recursive.
    public static ArrayList<Integer> inorderTraverse(TreeNode root) {
        ArrayList<Integer> all = new ArrayList<Integer>();
        inorder(root, all);
        return all;
    }

    public static void inorder(TreeNode node, ArrayList<Integer> all) {
        if (node != null) {
            inorder(node.left, all);
            all.add(node.val);
            inorder(node.right, all);
        }
    }
}


class TreeNode {
    TreeNode left;
    TreeNode right;
    int val;

    TreeNode(int x) {
		this.left = null;
		this.right = null;
        this.val = x;
    }
}

File: tree/leetcode_Binary_Tree_Inorder_Traversal.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode binary tree inorder traversal
 */

#include <stdio.h>
#include <vector>

using std::vector;


/**
 * Definition for a binary tree node.
 */
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
 public:
  vector<int> inorderTraversal(TreeNode* root) {
    vector<int> order;
    inorder(root, &order);
    return order;
  }

  void inorder(TreeNode* root, vector<int>* order) {
    if (root == NULL) {
      return;
    } else {
      inorder(root->left, order);
      order->push_back(root->val);
      inorder(root->right, order);
    }
  }
};


int main() {
  Solution sol;
  // Don't bother to test this because if it fails for real, I'd quit my job.
}
